daemon/src/rollover/maker.rs

```
@@ -0,0 +1,293 @@
+        n_payouts: usize,
+    ) -> Self {
+        Self {
+            tasks: Tasks::default(),
+            oracle_pk,
+            get_announcement,
+            n_payouts,
+            pending_protocols: HashMap::default(),
+            executor,
+        }
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle(&mut self, msg: NewInboundSubstream, ctx: &mut xtra::Context<Self>) {
+        let NewInboundSubstream { peer, stream } = msg; // TODO: Use `PeerId` for something!
+        let address = ctx.address().expect("we are alive");
+
+        self.tasks.add_fallible(
+            async move {
+                let mut framed =
+                    Framed::new(stream, JsonCodec::<ListenerMessage, DialerMessage>::new());
+
+                let propose = framed
+                    .next()
+                    .await
+                    .context("End of stream while receiving Propose")?
+                    .context("Failed to decode Propose")?
+                    .into_propose()?;
+
+                address
+                    .send(ProposeReceived {
+                        propose,
+                        framed,
+                        peer,
+                    })
+                    .await?;
+
+                anyhow::Ok(())
+            },
+            |e| async move { tracing::warn!("Failed to handle incoming rollover protocol: {e:#}") },
+        );
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: ProposeReceived) {
+        let ProposeReceived {
+            propose,
+            framed,
+            peer,
+        } = msg;
+        let order_id = propose.order_id;
+
+        if let Err(e) = self
+            .executor
+            .execute(order_id, |cfd| {
+                // TODO: Validate that the correct peer is invoking this?
```

Does it really matter if it's the wrong `PeerId`. It will never succeed if they don't have the correct secret keys.

We could fail early by checking here, but it's not a huge deal imo.

---

Yeah, this is just about failing early. A taker could "attack" a maker by sending bogus rollover proposals to keep the maker busy. No money would be lost, but it could result in a denial of service attack. Not super important yet but if we can easily do it it might be better to just do it now :)